### DISCUSSION TITLE
Can Stb_image lib support 32-bit png file?

### DISCUSSION THREAD

**Discussion by ouyangpeng** [nothings/stb#1342] [Q&A]
**Is your feature request related to a problem? Please describe.**
When I load a 3D model, some of the pictures are colored, which are 32-bit PNG, and the display is gray，like this:

**Describe the solution you'd like**

Then I use OpenCV to load these pictures of this 3D model, which becomes a normal color texture。

**Can stb_image lib support 32-bit png file?""
So my request is that in the futer , the image library stb_ image can support loading 32-bit PNG pictures,thanks!

**Comment by nothings** [comment]
PNG format only suports 1, 2, 4, 8, and 16 bits _per channel_, and allows 1, 2, 3, or 4 channels. By 32-bit image you probably mean 4 channels, 8-bits-per-channel. But it doesn't really matter, because stb_image supports all PNG formats. You are probably using the library incorretly.

**Reply by ouyangpeng** [reply]
> PNG format only suports 1, 2, 4, 8, and 16 bits _per channel_, and allows 1, 2, 3, or 4 channels. By 32-bit image you probably mean 4 channels, 8-bits-per-channel. But it doesn't really matter, because stb_image supports all PNG formats. You are probably using the library incorretly.


this is my code ,from the repo  https://github.com/ouyangpeng/OpenGLESDemo 

[ImageLoader.cpp](https://github.com/ouyangpeng/OpenGLESDemo/blob/master/app/src/main/cpp/utils/image/ImageLoader.cpp)
```cpp

#include "ImageLoader.h"

#define STB_IMAGE_IMPLEMENTATION
#include <stb/stb_image.h>
#include <GLUtils.h>

unsigned int
ImageLoader::TextureFromFile(const char *path, const std::string &directory, bool gamma) {
    std::string filename = std::string(path);
    filename = directory + '/' + filename;

    unsigned int textureID;
    glGenTextures(1, &textureID);

    // load the texture using stb_image
    LOGD("TextureFromFile Loading texture %s", filename.c_str())

    int width, height, nrComponents;
    unsigned char *data = stbi_load(filename.c_str(), &width, &height, &nrComponents, 0);
    if (data) {
        GLenum format;
        if (nrComponents == 1)
            format = GL_LUMINANCE;
        else if(nrComponents == 2)
            format = GL_LUMINANCE_ALPHA;
        else if (nrComponents == 3)
            format = GL_RGB;
        else if (nrComponents == 4)
            format = GL_RGBA;
        else
            format = GL_RGB;

        LOGD("nrComponents = %d",nrComponents)

        glBindTexture(GL_TEXTURE_2D, textureID);
        glTexImage2D(GL_TEXTURE_2D, 0, format, width, height, 0, format, GL_UNSIGNED_BYTE, data);
        glGenerateMipmap(GL_TEXTURE_2D);

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

        stbi_image_free(data);
    } else {
        LOGD("Texture failed to load at path: %s", path)
        stbi_image_free(data);
    }
    return textureID;
}
```

and load model using the above method, in [model.h](https://github.com/ouyangpeng/OpenGLESDemo/blob/master/app/src/main/cpp/model/model.h)

Some codes are as follows

```cpp
            if (!skip) {   // if texture hasn't been loaded already, load it
                Texture texture;
                // todo using stb_image lib show the model grey
                // texture.id = ImageLoader::TextureFromFile(str.C_Str(), this->directory);

                // using OpenCV to load the picture
                texture.id =TextureFromFile(str.C_Str(), this->directory);
                texture.type = typeName;
                texture.path = str.C_Str();
                textures.push_back(texture);
                
                // store it as texture loaded for entire model, to ensure we won't unnecesery load duplicate textures.
                textures_loaded.push_back(texture);
            }
```

Please help me to see if my imageloader.cpp using the stb_imge library incorretly

**Reply by nothings** [reply]
As long as your image width is a multiple of 4 or it's a 4-channel image, that looks correct at first glance.

Please post an example of an image that doesn't load correctly.

**Reply by ouyangpeng** [reply]
when i using stb_imge lib to  load the model's texure , like this:
![image](https://user-images.githubusercontent.com/5885735/177296986-67948a2d-030a-49cb-b492-828ded95f561.png)

and then i using opencv lib to load the model's texure , like this:
![image](https://user-images.githubusercontent.com/5885735/177297201-1b513918-bbd1-44da-b3f1-af674d65ea68.png)

the model link is [model source obj file link](https://github.com/ouyangpeng/OpenGLESDemo/tree/master/app/src/main/assets/poly) 。Please click the link

**Reply by nothings** [reply]
Apricot_02_diffuse.png loads correctly in my tests.

You're either using stb wrong, or you're accidentally loading one of the other images from that directory which is greyscale.

Have you looked at the "nrComponents =" line in your log file?

**Reply by ouyangpeng** [reply]
The contents of the log are as follows：
```cpp

2022-07-05 18:05:41.324 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][37]: nrComponents = 4
2022-07-05 18:05:41.579 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][37]: nrComponents = 4
2022-07-05 18:05:41.888 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][37]: nrComponents = 4

```

**Reply by ouyangpeng** [reply]
```
2022-07-05 18:07:16.567 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][20]: TextureFromFile Loading texture /data/data/com.oyp.openglesdemo/cache/poly/textures/Apricot_02_diffuse.png
2022-07-05 18:07:16.618 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][37]: nrComponents = 4
2022-07-05 18:07:16.619 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][20]: TextureFromFile Loading texture /data/data/com.oyp.openglesdemo/cache/poly/textures/Apricot_02_specular.png
2022-07-05 18:07:16.815 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][37]: nrComponents = 4
2022-07-05 18:07:16.825 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][20]: TextureFromFile Loading texture /data/data/com.oyp.openglesdemo/cache/poly/textures/Apricot_02_normal.png
2022-07-05 18:07:17.107 12114-12415/com.oyp.openglesdemo D/NDK_JNI_LOG_TAG: [ImageLoader.cpp][TextureFromFile][37]: nrComponents = 4
```

**Reply by nothings** [reply]
Anyway, I can't debug this issue for you. Nothing looks obviously wrong in your code, but I promise you, if in a trivial test app I get stb_image to load your apricot texture, then write it out as a BMP, the BMP I get is color.

Your image:
![Apricot_02_diffuse](https://user-images.githubusercontent.com/7525026/177304477-61f572cb-665c-4a46-a356-756671f9cb12.png)

Loaded by stb_image then written out as BMP then converted to JPG (github doesn't support BMP):
![apricot_02_diffuse](https://user-images.githubusercontent.com/7525026/177305085-9206b63b-0f88-4bc3-a0fa-7bf882b0d4aa.jpg)

**Reply by nothings** [reply]
Let me put it this way: call stbi_load with a known color image, and look at the actual values returned from it (in a logfile, or in the debugger). If it's grey, there will be groups of three identical bytes, like 200,200,200, 150,150,150, 99,99,99. If it's color, there won't be that simple pattern.

I guarantee you that stb_image on apricot_02_diffuse.png is not producing triples like that. But you're not looking at the output of stb_image, you're looking at a whole pipeline of stuff operating between stb_image's output and the screen. stb_image's job is just to not produce those identical triples, and I'm sure it's not. If you can show it *is* producing identical triples, then there is something for me to look at. But if it's not producing identical triples, the problem is on your side of the fence, it's got nothing to do with me.

**Reply by ouyangpeng** [reply]
thanks   i use stb_image.h to load this pic, and then use stb_image_write.h and stb_image_resize.h to resize the pic and store

```cpp
#include <iostream>

#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"
#define STB_IMAGE_RESIZE_IMPLEMENTATION
#include "stb_image_resize.h"
#include <string>
#include <stdio.h>
#include <stdlib.h>
#include <vector>

using namespace std;

int main() {
	std::cout << "Hello, STB_Image" << std::endl;

	string inputPath = "./Apricot_02_diffuse.png";
	int imageWidth, imageHeight, nrComponents;

	// 加载图片获取宽、高、颜色通道信息
	unsigned char *imagedata = stbi_load(inputPath.c_str(), &imageWidth, &imageHeight, &nrComponents, STBI_default);
    std::cout << "测试 STB_Image 读取到的信息为：imageWidth = "<< imageWidth << 
             ",imageHeight = "<<imageHeight << ", nrComponents = " << nrComponents 
             << std::endl;  

    // 输出的宽和高为原来的一半
	int outputWidth = imageWidth / 2;
	int outputHeight = imageHeight / 2;
    // 申请内存
	auto *outputImageData = (unsigned char *)malloc(outputWidth * outputHeight * nrComponents);
    std::cout << "测试 STBIR_RESIZE, 改变后的图片尺寸为：outputWidth = "<< outputWidth<<", outputHeight = "<< outputHeight << std::endl;
	// 改变图片尺寸
	stbir_resize(imagedata, imageWidth, imageHeight, 0, outputImageData, outputWidth, outputHeight, 0, STBIR_TYPE_UINT8, nrComponents, STBIR_ALPHA_CHANNEL_NONE, 0,
		STBIR_EDGE_CLAMP, STBIR_EDGE_CLAMP,
		STBIR_FILTER_BOX, STBIR_FILTER_BOX,
		STBIR_COLORSPACE_SRGB, nullptr
		);

	string outputPath = "./Apricot_02_diffuse_out.png";
    std::cout << "测试 STB_WRITE 要写入的地址为： "<< outputPath << std::endl;
	// 写入图片
	stbi_write_png(outputPath.c_str(), outputWidth, outputHeight, nrComponents, outputImageData, 0);

	stbi_image_free(imagedata);
	stbi_image_free(outputImageData);
	return 0;
}

```

the original pic is :

![image](https://user-images.githubusercontent.com/5885735/177305883-62783edc-4d4d-4a39-ba78-f0cd4ad0965e.png)

after resize , the pic is :
![image](https://user-images.githubusercontent.com/5885735/177305911-9987aefa-273c-4ed3-a265-f8835b677a7a.png)


i can't find where the wrong when i load the 3d model , oh my god

**Reply by nothings** [reply]
Good luck, too bad there's nothing obviously wrong!

**Reply by ouyangpeng** [reply]
OK, I'll check what causes the gray display of the loaded 3D model. I'll reply when I find it. Thank you very much!
